src/persist-txn/src/txn_cache.rs

```diff
@@ -809,51 +855,65 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64, C: TxnsCodec>
     /// Invariant: afterward, self.progress_exclusive will be > ts
     #[instrument(level = "debug", fields(ts = ?ts))]
     pub async fn update_gt(&mut self, ts: &T) {
-        self.update(|progress_exclusive| progress_exclusive > ts)
-            .await;
+        let only_data_id = self.only_data_id.clone();
+        Self::update(
+            &mut self.state,
+            &mut self.txns_subscribe,
+            &mut self.buf,
+            only_data_id,
+            |progress_exclusive| progress_exclusive > ts,
+        )
+        .await;
         debug_assert!(&self.progress_exclusive > ts);
         debug_assert_eq!(self.validate(), Ok(()));
     }
 
     /// Invariant: afterward, self.progress_exclusive will be >= ts
     #[instrument(level = "debug", fields(ts = ?ts))]
     pub async fn update_ge(&mut self, ts: &T) {
-        self.update(|progress_exclusive| progress_exclusive >= ts)
-            .await;
+        let only_data_id = self.only_data_id.clone();
+        Self::update(
+            &mut self.state,
+            &mut self.txns_subscribe,
+            &mut self.buf,
+            only_data_id,
+            |progress_exclusive| progress_exclusive >= ts,
+        )
+        .await;
         debug_assert!(&self.progress_exclusive >= ts);
         debug_assert_eq!(self.validate(), Ok(()));
     }
 
-    async fn update<F: Fn(&T) -> bool>(&mut self, done: F) {
-        while !done(&self.progress_exclusive) {
-            let events = self.txns_subscribe.next(None).await;
+    async fn update<F: Fn(&T) -> bool>(
+        state: &mut TxnsCacheState<T>,
+        txns_subscribe: &mut Subscribe<C::Key, C::Val, T, i64>,
+        buf: &mut Vec<(TxnsEntry, T, i64)>,
+        only_data_id: Option<ShardId>,
```

Comment:
Technically I could make this `Cow<Option<ShardId>>` since we only need an owned version sometimes, but that seemed like overkill.

Reply:
`ShardId` is `Copy` anyway :)